Added Storing And Loading Saved Layouts

Eroxl opened the pull request:
Added the ability to store and load saved layouts from 4 prefabs slots

Closes #248

Eroxl committed:
Added Preset Buttons

Added 4 preset buttons to save and load presets for this suggestion https://github.com/LiveTL/LiveTL/issues/248

Eroxl committed:
Added Preset Buttons

Added 4 preset buttons to save and load presets for this suggestion https://github.com/LiveTL/LiveTL/issues/248

Eroxl committed:
Changed Function From Get Entire To Get

Eroxl committed:
Changed Function From Get Entire To Get

Eroxl committed:
Changed The Template To Use A For Loop

Cleaned up the template so that it uses a for loop instead of an explicit template, could still be improved by generating an array but I couldn't find a very clean way to do it.

Eroxl committed:
Changed The Template To Use A For Loop

Cleaned up the template so that it uses a for loop instead of an explicit template, could still be improved by generating an array but I couldn't find a very clean way to do it.

Eroxl committed:
Removed Enable Full Screen Button From Presets

Eroxl committed:
Removed Enable Full Screen Button From Presets

Eroxl committed:
Removed Preset Options

Removed some preset options that don't apply to the formatting and don't need to be saved into a preset.

Eroxl committed:
Removed Preset Options

Removed some preset options that don't apply to the formatting and don't need to be saved into a preset.

Eroxl committed:
Formatted File With ESLint

Formatted the file and fixed the errors.

Eroxl committed:
Formatted File With ESLint

Formatted the file and fixed the errors.

KentoNishi commented:
yoooooooooo wait this is sick, i'll review it when i have time

r2dev2 commented on src/components/settings/PresetButtons.svelte:
```diff
@@ -0,0 +1,56 @@
+  import { importStores } from '../../js/storage';
+  import Button from 'smelte/src/components/Button';
+
+  $: active = $activePreset;
+  $: activePreset.set(active);
```
Is it needed to bind the variable to the store this way for this component? I think we only used this technique in components in which we had to two-way bind to a store's value which sometimes doesn't work.

r2dev2 commented on src/components/settings/PresetButtons.svelte:
```diff
@@ -0,0 +1,56 @@
+  $: active = $activePreset;
+  $: activePreset.set(active);
+
+  /**
+  * @param {number} number
+  */
```
We started using typescript instead of jsdoc, can you change this to a ts hint and make the script tag a `<script lang="ts">`?

r2dev2 commented on src/components/settings/PresetButtons.svelte:
```diff
@@ -0,0 +1,56 @@
+      return;
+    }
+
+    const presetData = presets.get();
+    presetData[number - 1] = {
+      captionDuration: captionDuration.get(),
+      captionFontSize: captionFontSize.get(),
+      videoSideSetting: videoSideSetting.get(),
+      chatZoom: chatZoom.get(),
+      livetlFontSize: livetlFontSize.get(),
+      textDirection: textDirection.get(),
+      chatSplit: chatSplit.get(),
+      autoVertical: autoVertical.get(),
+      autoLaunchMode: autoLaunchMode.get()
```
You can use `$store` instead of `store.get()` here (and line 26 as well). 

Can you also save the width and height of the splits in addition to the keyboard shortcuts and translator mode configurations in these settings?

KentoNishi reviewed (changes requested):
Aside from the implementational details @r2dev2 pointed out, I think conceptually, the limit of 4 layouts is pretty arbitrary. I would prefer there to be a customizable amount where a user can freely add or remove presets

Eroxl committed:
Fixed Suggestions

Changed the language tag to typescript and switched the types to typescript.

Removed the unnecessary double binding of the active variable

Removed the .get() functions and replaced them with $'s

And added translator mode settings, key-bind's and the window size to the presets as well as slightly changed the formatting of the code.

Eroxl committed:
Fixed Suggestions

Changed the language tag to typescript and switched the types to typescript.

Removed the unnecessary double binding of the active variable

Removed the .get() functions and replaced them with $'s

And added translator mode settings, key-bind's and the window size to the presets as well as slightly changed the formatting of the code.

Eroxl committed:
Updated Presets To Work With An Arbitrarily Large Amount Of Presets

Updated the presets to work with any amount of presets the user wants and added two new buttons to add and remove presets.

Eroxl committed:
Updated Presets To Work With An Arbitrarily Large Amount Of Presets

Updated the presets to work with any amount of presets the user wants and added two new buttons to add and remove presets.

Eroxl commented:
Check it out again I made the main changes. 

The code for adding and removing presets is pretty ugly but it works reliably.

KentoNishi commented on src/components/settings/PresetButtons.svelte:
```diff
@@ -0,0 +1,136 @@
+{#each getPresetArray(presetCount) as i} 
+  <div class="flex gap-2 py-1">
+    <Button add="flex-1" on:click={() => { isDeleting ? deletePreset(2 * i - 1) : updatePreset(2 * i - 1); }} color="dark" light={active === 2 * i - 1 && !isDeleting}>
+      {isDeleting ? `Delete Preset ${2 * i - 1}` : active !== 2 * i - 1 ? `Preset ${2 * i - 1}` : `Save Preset ${2 * i - 1}`}
```
oh god, can you refactor this so it's more clear? otherwise i'll have a look at it myself when i have time (which might take a while sorry kinda of busy recently)

r2dev2 commented on src/components/settings/PresetButtons.svelte:
```diff
@@ -0,0 +1,136 @@
+
+{#each getPresetArray(presetCount) as i} 
+  <div class="flex gap-2 py-1">
+    <Button add="flex-1" on:click={() => { isDeleting ? deletePreset(2 * i - 1) : updatePreset(2 * i - 1); }} color="dark" light={active === 2 * i - 1 && !isDeleting}>
+      {isDeleting ? `Delete Preset ${2 * i - 1}` : active !== 2 * i - 1 ? `Preset ${2 * i - 1}` : `Save Preset ${2 * i - 1}`}
+    </Button>
+    {#if i !== presetCount / 2 + 0.5}
+      <Button add="flex-1" on:click={() => { isDeleting ? deletePreset(2 * i) : updatePreset(2 * i); }} color="dark" light={active === 2 * i && !isDeleting}>
+        {isDeleting ? `Delete Preset ${2 * i}` : active !== 2 * i ? `Preset ${2 * i}` : `Save Preset ${2 * i}`}
+      </Button>
+    {/if}
```
Instead of putting the buttons next to each other in js, you should make them next to each other with css. Like if the screen isn't wide, they can be on top of each other but as the space widens, the buttons start to form rows.

I have an example on [my website (change the window width and observe the project cards)](https://r2dev2.github.io/) with the css to do the effect [in the source code](https://github.com/r2dev2/r2dev2.github.io/blob/d7df0489e03276443dfb2e4ce8009a12719ee10b/src/App.svelte#L78). Doing something similar but with flex display could get you a bottom row that is completely full.

Eroxl replied:
I was gonna do that but I when i squished the screen past a certain point no other components shrunk like the import/export button so I never implemented it so that they would keep the same style as the import/export buttons.

Eroxl replied:
I'm just gonna switch it over from flexbox to a css grid quickly so that it cleans up the code so I don't need to explicitly do even and odd numbers for the time being and then I'll start working on automatically switching them over to a vertical mode.

Eroxl replied:
I think the whole UI needs to be redesigned so that it works with smaller screen sizes before I can do that because it overflows right now.

r2dev2 commented on src/components/settings/PresetButtons.svelte:
```diff
@@ -0,0 +1,136 @@
+    }
+
+    const presetData = $presets;
+    presetData[number - 1] = {
+      captionDuration: $captionDuration,
+      captionFontSize: $captionFontSize,
+      captionWidth: $captionWidth,
+      captionLeft: $captionLeft,
+      captionTop: $captionTop,
+
+      videoSideSetting: $videoSideSetting,
+      videoPanelSize: $videoPanelSize,
+
+      chatZoom: $chatZoom,
+      chatSize: $chatSize,
+      chatSplit: $chatSplit,
+
+      textDirection: $textDirection,
+      livetlFontSize: $livetlFontSize,
+
+      keyboardShortcuts: $keyboardShortcuts,
+
+      doTranslatorMode: $doTranslatorMode,
+      doAutoPrefix: $doAutoPrefix,
+      autoPrefixTag: $autoPrefixTag,
+      macroTrigger: $macroTrigger,
+      macros: $macros
```
There seem to be a lot of stores here in this massive object. It might be better to export an array of these stores from `store.js` and then do
```js
// presetStores is an array with all the stores that will be persisted in presets
presetData[number - 1] = Object.fromEntries(presetStores.map(store => [store.name, store.get()]));
```
and do something similar for the default value object you make later.

Eroxl committed:
Clead Up Preset Button Code

Cleaned up the preset button code by moving the button code into it's own script and cleaning up the way the variables are generated.

Eroxl committed:
Clead Up Preset Button Code

Cleaned up the preset button code by moving the button code into it's own script and cleaning up the way the variables are generated.

Eroxl committed:
Changed Variable Names

Changed variables names so that they're a little more readable / explicative.

Eroxl committed:
Changed Variable Names

Changed variables names so that they're a little more readable / explicative.

Eroxl committed:
Changed Variable Names

Changed the odd and even variable names to oddNumber and evenNumber so the if statement makes them look less like booleans.

Eroxl committed:
Changed Variable Names

Changed the odd and even variable names to oddNumber and evenNumber so the if statement makes them look less like booleans.

Eroxl committed:
Moved Large List Of Variables Into Store Array 

Moved the large list of persistent variables into an array in store.js.

Eroxl committed:
Moved Large List Of Variables Into Store Array 

Moved the large list of persistent variables into an array in store.js.

Eroxl committed:
Removed Unnecessary Parts Of The Presets Array

Eroxl committed:
Removed Unnecessary Parts Of The Presets Array

Eroxl committed:
Changed Preset Button From Flex Box To CSS Grid.

Changed the preset button from flex box to css grid so that it can be more dynamically generated. 

Also formatted the file.

Eroxl committed:
Changed Preset Button From Flex Box To CSS Grid.

Changed the preset button from flex box to css grid so that it can be more dynamically generated. 

Also formatted the file.

Eroxl committed:
Removed Unnecessary CSS

Eroxl committed:
Removed Unnecessary CSS

Eroxl committed:
Fixed HC Submodule Version

Bumped the chat version that got downgraded.

Eroxl committed:
Fixed Only Emoji's Issue

r2dev2 committed:
Merge pull request #357 from Eroxl/only-emojis-fix

KentoNishi commented on src/components/settings/PresetButtons.svelte:
```diff
@@ -0,0 +1,66 @@
+<div class="grid grid-cols-2 gap-2 py-1">
+  {#each Array(presetAmount) as _, i}
+    <PresetButton 
+      activeNumber = { activeNumber }
```
two things here
1. you can do `{ activeNumber }`
2. you probably should use event dispatchers instead (basically event listeners) and listen to `on:eventName`. Check out https://svelte.dev/tutorial/component-events

Eroxl replied:
Ok should be fixed now.

KentoNishi commented on src/components/settings/PresetButtons.svelte:
```diff
@@ -0,0 +1,66 @@
+  </Button>
+  <Button color="error" on:click={() => { isDeleting = !isDeleting; }}>
+    {
+      isDeleting ? 'Stop Deleting' : 'Delete Presets'
```
awkward formatting, i'll just run my svelte autoformatter config before merge so don't worry about it for now

Eroxl committed:
Switch To Event Listeners

Switched the preset buttons to use event listeners instead of explicit functions.

Eroxl committed:
Switch To Event Listeners

Switched the preset buttons to use event listeners instead of explicit functions.

r2dev2 reviewed (approved):
Formatting is a bit weird but I'm assuming @KentoNishi will run the autoformatter on this.

Eroxl commented:
@r2dev2 Yeah https://github.com/LiveTL/LiveTL/pull/356#discussion_r743031772
I just ran the auto eslint formatter with the npm command.

Eroxl committed:
Merge remote-tracking branch 'origin/develop' into develop

Eroxl commented:
Just had to re-base but I didn't want to downgrade the Hyper Chat version again.

KentoNishi committed:
formatting + ui tweaks

KentoNishi reviewed (changes requested):
One more thing I would love for this feature: naming presets! I think it'd be really nice touch, could you try adding that?

Eroxl commented:
@KentoNishi Ok I'll see about a good way to try adding it.

Eroxl committed:
Added Renaming Presets

Added the ability to rename presets through a new green button with the edit icon.

Eroxl committed:
Added Auto Focus To Renaming

Added auto focus to the renaming section of the preset buttons so it's a bit more clear to the user that they can type and rename the preset.

Eroxl committed:
Made The Enter Key Close And Save The Preset Name

Made the enter key close and save the preset so it's more similar to accessibility standards.

Eroxl committed:
Removed Temporary Variables

Removed some temporary variables I was using to test out a way to only allow one to be editable at a time.

Eroxl committed:
Made Only One Preset Name Editable At A Time

Made it so that only one preset name can be edited at a time so that it's a bit less confusing for users.

Eroxl commented:
@KentoNishi Check it out now I added the renaming of presets.

KentoNishi commented:
yooo pog, will check it out when i have time

KentoNishi commented:
![image](https://user-images.githubusercontent.com/38841491/141255519-df3538fa-ce9d-4ffd-a810-69a261d47c3b.png)
something is not right here lol

Eroxl commented:
@KentoNishi Definitely not

KentoNishi committed:
formatting, @r2dev2 do the jsdoc plz

Eroxl commented:
@KentoNishi Mind exporting your settings so I can take a look?

KentoNishi commented:
`{"language":"English","showModMessage":true,"chatZoom":1,"showTimestamp":true,"textDirection":"BOTTOM","videoSide":"LEFT","showCaption":true,"captionDuration":10,"captionFontSize":18,"captionWidth":80,"captionLeft":10,"captionTop":80,"doSpeechSynth":true,"speechVolume":1,"doTranslatorMode":false,"videoPanelSize":49.21875,"chatSize":50.0515995872033,"livetlFontSize":18,"textFilters":[],"textBlacklist":[],"plaintextWhitelist":[],"plaintextBlacklist":[],"userFilters":{},"channelFilters":{},"plainAuthorWhitelist":[],"regexAuthorWhitelist":[],"plainAuthorBlacklist":[],"regexAuthorBlacklist":[],"customFilters":[],"enableCaptionTimeout":false,"chatSplit":"HORIZONTAL","lastVersion":"","screenshotRenderWidth":500,"welcomeDismissed":false,"macros":[],"doAutoPrefix":false,"enableMchadTLs":true,"enableAPITLs":true,"enableExportButtons":true,"enableFullscreenButton":true,"mchadUsers":{},"autoPrefixTag":"[$filterLang]","macroTrigger":"/","ytcDeleteBehaviour":"HIDE","autoVertical":true,"enableSpamProtection":true,"spamMsgAmount":5,"spamMsgInterval":10,"spammersDetected":{},"speechVoiceNameSetting":"Microsoft David - English (United States)","speechSpeed":1,"autoLaunchMode":"NONE","keyboardShortcuts":{"fullScreen":"f","toggleMute":"m","togglePlayPause":"<Space>","volumeDown":"ArrowDown","volumeUp":"ArrowUp"},"disableSpecialSpamProtection":true,"activePreset":4,"presets":[{"autoPrefixTag":"[$filterLang]","captionDuration":10,"captionFontSize":18,"captionLeft":3.4574468085106385,"captionTop":31.614349775784753,"captionWidth":79.98670212765957,"chatSize":43.60986547085202,"chatSplit":"HORIZONTAL","chatZoom":1,"doAutoPrefix":false,"doTranslatorMode":false,"keyboardShortcuts":{"fullScreen":"f","toggleMute":"m","togglePlayPause":"<Space>","volumeDown":"ArrowDown","volumeUp":"ArrowUp"},"livetlFontSize":18,"macroTrigger":"/","macros":[],"textDirection":"BOTTOM","videoPanelSize":61.369680851063826,"videoSide":"LEFT"},{"autoPrefixTag":"[$filterLang]","captionDuration":10,"captionFontSize":18,"captionLeft":3.4574468085106385,"captionTop":31.614349775784753,"captionWidth":79.98670212765957,"chatSize":43.64303178484108,"chatSplit":"HORIZONTAL","chatZoom":1,"doAutoPrefix":false,"doTranslatorMode":false,"keyboardShortcuts":{"fullScreen":"f","toggleMute":"m","togglePlayPause":"<Space>","volumeDown":"ArrowDown","volumeUp":"ArrowUp"},"livetlFontSize":18,"macroTrigger":"/","macros":[],"textDirection":"BOTTOM","videoPanelSize":14.791666666666666,"videoSideSetting":"LEFT","name":"test"},{"captionDuration":10,"captionFontSize":18,"captionWidth":79.98670212765957,"captionLeft":3.4574468085106385,"captionTop":31.614349775784753,"videoSide":"LEFT","chatZoom":1,"livetlFontSize":18,"textDirection":"BOTTOM","chatSplit":"HORIZONTAL","videoPanelSize":14.827127659574469,"chatSize":63.11659192825112,"keyboardShortcuts":{"fullScreen":"f","toggleMute":"m","togglePlayPause":"<Space>","volumeDown":"ArrowDown","volumeUp":"ArrowUp"},"doTranslatorMode":false,"doAutoPrefix":false,"autoPrefixTag":"[$filterLang]","macroTrigger":"/","macros":[]},{"autoPrefixTag":"[$filterLang]","captionDuration":10,"captionFontSize":18,"captionLeft":10,"captionTop":80,"captionWidth":80,"chatSize":50,"chatSplit":"HORIZONTAL","chatZoom":1,"doAutoPrefix":false,"doTranslatorMode":false,"keyboardShortcuts":{"fullScreen":"f","toggleMute":"m","togglePlayPause":"<Space>","volumeDown":"ArrowDown","volumeUp":"ArrowUp"},"livetlFontSize":18,"macroTrigger":"/","macros":[],"textDirection":"BOTTOM","videoPanelSize":70,"videoSide":"LEFT"},{"autoPrefixTag":"[$filterLang]","captionDuration":10,"captionFontSize":18,"captionLeft":7.912234042553192,"captionTop":44.84304932735426,"captionWidth":79.98670212765957,"chatSize":50.0515995872033,"chatSplit":"HORIZONTAL","chatZoom":1,"doAutoPrefix":false,"doTranslatorMode":false,"keyboardShortcuts":{"fullScreen":"f","toggleMute":"m","togglePlayPause":"<Space>","volumeDown":"ArrowDown","volumeUp":"ArrowUp"},"livetlFontSize":18,"macroTrigger":"/","macros":[],"textDirection":"BOTTOM","videoPanelSize":53.489583333333336,"videoSide":"LEFT"},{"captionDuration":10,"captionFontSize":18,"captionWidth":80,"captionLeft":10,"captionTop":80,"videoSide":"LEFT","chatZoom":1,"livetlFontSize":18,"textDirection":"BOTTOM","chatSplit":"HORIZONTAL","videoPanelSize":70,"chatSize":50,"keyboardShortcuts":{"volumeUp":"ArrowUp","volumeDown":"ArrowDown","fullScreen":"f","toggleMute":"m","togglePlayPause":"<Space>"},"doTranslatorMode":false,"doAutoPrefix":false,"autoPrefixTag":"[$filterLang]","macroTrigger":"/","macros":[],"name":"Preset"},{"captionDuration":10,"captionFontSize":18,"captionWidth":80,"captionLeft":10,"captionTop":80,"videoSide":"LEFT","chatZoom":1,"livetlFontSize":18,"textDirection":"BOTTOM","chatSplit":"HORIZONTAL","videoPanelSize":70,"chatSize":50,"keyboardShortcuts":{"volumeUp":"ArrowUp","volumeDown":"ArrowDown","fullScreen":"f","toggleMute":"m","togglePlayPause":"<Space>"},"doTranslatorMode":false,"doAutoPrefix":false,"autoPrefixTag":"[$filterLang]","macroTrigger":"/","macros":[],"name":"Preset"}]}`

Eroxl commented:
Thanks

KentoNishi commented:
there's also a lot of typing issues like the presets store not being typed, i'll let @r2dev2 handle that after ur done

Eroxl commented:
Ok I think it might just be an issue with upgrading from the pre-naming version to the naming version because some of your presets don't have names. That shouldn't be able to happen in production.

But I also found another bug with importing presets where they don't render till you reload so I'm gonna try to fix that quickly.

I'll post a comment with a config that should fix your file next just with the names added.

Eroxl commented:
{"language":"English","showModMessage":true,"chatZoom":1,"showTimestamp":true,"textDirection":"BOTTOM","videoSide":"LEFT","showCaption":true,"captionDuration":10,"captionFontSize":18,"captionWidth":80,"captionLeft":10,"captionTop":80,"doSpeechSynth":true,"speechVolume":1,"doTranslatorMode":false,"videoPanelSize":70,"chatSize":50,"livetlFontSize":18,"textFilters":[],"textBlacklist":[],"plaintextWhitelist":[],"plaintextBlacklist":[],"userFilters":{},"channelFilters":{},"plainAuthorWhitelist":[],"regexAuthorWhitelist":[],"plainAuthorBlacklist":[],"regexAuthorBlacklist":[],"customFilters":[],"enableCaptionTimeout":false,"chatSplit":"HORIZONTAL","lastVersion":"","screenshotRenderWidth":500,"welcomeDismissed":false,"macros":[],"doAutoPrefix":false,"enableMchadTLs":true,"enableAPITLs":true,"enableExportButtons":true,"enableFullscreenButton":true,"mchadUsers":{},"autoPrefixTag":"[$filterLang]","macroTrigger":"/","ytcDeleteBehaviour":"HIDE","autoVertical":true,"enableSpamProtection":true,"spamMsgAmount":5,"spamMsgInterval":10,"spammersDetected":{},"speechVoiceNameSetting":"Alex","speechSpeed":1,"autoLaunchMode":"NONE","keyboardShortcuts":{"volumeUp":"ArrowUp","volumeDown":"ArrowDown","fullScreen":"f","toggleMute":"m","togglePlayPause":"<Space>"},"disableSpecialSpamProtection":true,"activePreset":7,"presets":[{"captionDuration":10,"captionFontSize":18,"captionWidth":79.98670212765957,"captionLeft":3.4574468085106385,"captionTop":31.614349775784753,"videoSide":"LEFT","chatZoom":1,"livetlFontSize":18,"textDirection":"BOTTOM","chatSplit":"HORIZONTAL","videoPanelSize":61.369680851063826,"chatSize":43.60986547085202,"keyboardShortcuts":{"fullScreen":"f","toggleMute":"m","togglePlayPause":"<Space>","volumeDown":"ArrowDown","volumeUp":"ArrowUp"},"doTranslatorMode":false,"doAutoPrefix":false,"autoPrefixTag":"[$filterLang]","macroTrigger":"/","macros":[],"name":"Preset"},{"captionDuration":10,"captionFontSize":18,"captionWidth":79.98670212765957,"captionLeft":3.4574468085106385,"captionTop":31.614349775784753,"videoSide":"LEFT","chatZoom":1,"livetlFontSize":18,"textDirection":"BOTTOM","chatSplit":"HORIZONTAL","videoPanelSize":14.791666666666666,"chatSize":43.64303178484108,"keyboardShortcuts":{"fullScreen":"f","toggleMute":"m","togglePlayPause":"<Space>","volumeDown":"ArrowDown","volumeUp":"ArrowUp"},"doTranslatorMode":false,"doAutoPrefix":false,"autoPrefixTag":"[$filterLang]","macroTrigger":"/","macros":[],"name":"Preset"},{"captionDuration":10,"captionFontSize":18,"captionWidth":79.98670212765957,"captionLeft":3.4574468085106385,"captionTop":31.614349775784753,"videoSide":"LEFT","chatZoom":1,"livetlFontSize":18,"textDirection":"BOTTOM","chatSplit":"HORIZONTAL","videoPanelSize":14.827127659574469,"chatSize":63.11659192825112,"keyboardShortcuts":{"fullScreen":"f","toggleMute":"m","togglePlayPause":"<Space>","volumeDown":"ArrowDown","volumeUp":"ArrowUp"},"doTranslatorMode":false,"doAutoPrefix":false,"autoPrefixTag":"[$filterLang]","macroTrigger":"/","macros":[],"name":"Preset"},{"captionDuration":10,"captionFontSize":18,"captionWidth":80,"captionLeft":10,"captionTop":80,"videoSide":"LEFT","chatZoom":1,"livetlFontSize":18,"textDirection":"BOTTOM","chatSplit":"HORIZONTAL","videoPanelSize":70,"chatSize":50,"keyboardShortcuts":{"fullScreen":"f","toggleMute":"m","togglePlayPause":"<Space>","volumeDown":"ArrowDown","volumeUp":"ArrowUp"},"doTranslatorMode":false,"doAutoPrefix":false,"autoPrefixTag":"[$filterLang]","macroTrigger":"/","macros":[],"name":"test"},{"captionDuration":10,"captionFontSize":18,"captionWidth":79.98670212765957,"captionLeft":7.912234042553192,"captionTop":44.84304932735426,"videoSide":"LEFT","chatZoom":1,"livetlFontSize":18,"textDirection":"BOTTOM","chatSplit":"HORIZONTAL","videoPanelSize":53.489583333333336,"chatSize":50.0515995872033,"keyboardShortcuts":{"fullScreen":"f","toggleMute":"m","togglePlayPause":"<Space>","volumeDown":"ArrowDown","volumeUp":"ArrowUp"},"doTranslatorMode":false,"doAutoPrefix":false,"autoPrefixTag":"[$filterLang]","macroTrigger":"/","macros":[],"name":"Preset"},{"captionDuration":10,"captionFontSize":18,"captionWidth":80,"captionLeft":10,"captionTop":80,"videoSide":"LEFT","chatZoom":1,"livetlFontSize":18,"textDirection":"BOTTOM","chatSplit":"HORIZONTAL","videoPanelSize":70,"chatSize":50,"keyboardShortcuts":{"volumeUp":"ArrowUp","volumeDown":"ArrowDown","fullScreen":"f","toggleMute":"m","togglePlayPause":"<Space>"},"doTranslatorMode":false,"doAutoPrefix":false,"autoPrefixTag":"[$filterLang]","macroTrigger":"/","macros":[],"name":"Preset"},{"captionDuration":10,"captionFontSize":18,"captionWidth":80,"captionLeft":10,"captionTop":80,"videoSide":"LEFT","chatZoom":1,"livetlFontSize":18,"textDirection":"BOTTOM","chatSplit":"HORIZONTAL","videoPanelSize":70,"chatSize":50,"keyboardShortcuts":{"volumeUp":"ArrowUp","volumeDown":"ArrowDown","fullScreen":"f","toggleMute":"m","togglePlayPause":"<Space>"},"doTranslatorMode":false,"doAutoPrefix":false,"autoPrefixTag":"[$filterLang]","macroTrigger":"/","macros":[],"name":"Preset"}]}

Eroxl commented:
@KentoNishi Actually I'm probably just going to have to fix the rendering issue tomorrow because it's tied into the way I store the preset count.

r2dev2 committed:
type the presets store

KentoNishi commented:
@Eroxl sorry for the delayed review in the first place :|

there's no rush so take your time, lemme know when it's ready for review again :)

Eroxl committed:
Fix For Importing Bug

Fix for the importing that was causing it to not render because the data wasn't being updated on import.

Eroxl commented:
Nevermind the fix was really simple should be all good now.

KentoNishi commented:
i noticed that you do `let variable: type = defaultValue;` a lot. TS will automatically know the type if u give it a default value. i won't bother changing it, but just fyi

KentoNishi committed:
autofixes

KentoNishi committed:
small textual tweaks

r2dev2 commented on src/components/settings/PresetButton.svelte:
```diff
@@ -0,0 +1,78 @@
+
+  currentlyEditingPreset.subscribe((value) => {
+    if (value !== prefabNumber) isEditing = false;
+  });
```
you can do
```js
$: if ($currentlyEditingPreset !== prefabNumber) isEditing = false;
```

r2dev2 commented on src/components/settings/PresetButton.svelte:
```diff
@@ -0,0 +1,78 @@
+      if (isEditing) currentlyEditingPreset.set(prefabNumber);
+    }}
+  >
+    <Icon>{'edit'}</Icon>
```
you can just do `<Icon>edit</Icon>`

r2dev2 reviewed (changes requested):
Minor things, but otherwise looks good.

Eroxl committed:
r2dev2's Requested Changes

Fixed the changes r2dev2 requested.

Eroxl committed:
Removed Unnecessary TypeScript Types.

Eroxl committed:
Merge remote-tracking branch 'origin/develop' into develop

ChrRubin reviewed (changes requested):
The first preset does not have a default name, so saving it or adding a new preset without renaming the first preset will cause it to show as undefined:

![image](https://user-images.githubusercontent.com/20059164/142351384-f0652197-ccb7-4d7d-9164-0d0496352728.png)

Also I would suggest changing the preset buttons' colors to red, as well as hiding the "Add Preset" button and each edit button when `isDeleting` for a slightly better UX.

Eroxl committed:
Fixed First Prefab Being Undefined

Fixed the first prefab being undefined it was just a temporary fix for typing i forgot to remove.

Eroxl commented:
@ChrRubin The changes look good besides changing the preset buttons colours to red I think it's just too many bright colours and it gets hard to read
### With The Red
<img width="420" alt="Screen Shot 2021-11-17 at 9 27 21 PM" src="https://user-images.githubusercontent.com/77178777/142357221-8deef90e-6193-4f36-9296-970180eb86fc.png">

### Without The Red
<img width="421" alt="Screen Shot 2021-11-17 at 9 27 29 PM" src="https://user-images.githubusercontent.com/77178777/142357223-f8b4a407-67e3-4f45-bcb2-3a6d93ec8961.png">

Eroxl committed:
Style Improvements

ChrRubin commented:
Yeah I agree the red looks too bright
